### ISSUE TITLE
Cannot push, using ssh

### ISSUE THREAD

**Issue by dncohen** [gogits/gogs#3395]
- Gogs version (or commit ref): 
commit 08c976f8117b92b25ea197e7dc0372b49d9ee412
Author: Unknwon <u@gogs.io>
Date:   Sat Aug 6 11:20:10 2016 -0700

    Only do go vet on this codebase

- Git version: 
$ git --version
git version 2.7.0

- Operating system: 
OpenBSD 5.9

- Database (use `[x]`):
  - [X ] PostgreSQL
  - [ ] MySQL
  - [ ] SQLite
(Wanted to try sqlite, but couldn't make it work, not what this issue is about)

- Can you reproduce the bug at https://try.gogs.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [X ] Not relevant
- Log gist:

## Description

I'm aware that other issues in this queue have basically the same title.  I'm seeing different details, so I thought it wise to submit a new issue.  Apologies in advance if I'm wrong about that.

Here's what it looks like, to the user attempting to push:

```
dave@tripper:test $ git push origin 
Counting objects: 3, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (2/2), done.
Writing objects: 100% (3/3), 268 bytes | 0 bytes/s, done.
Total 3 (delta 0), reused 0 (delta 0)
remote: 2016/08/08 00:55:16 [W] Custom config 'custom/conf/app.ini' not found, ignore this if you're running first time
remote: error: hook declined to update refs/heads/master
To ssh://gogs@gogs.electricgroups.com:3022/dave/test.git
 ! [remote rejected] master -> master (hook declined)
error: failed to push some refs to 'ssh://gogs@gogs.electricgroups.com:3022/dave/test.git'
```

(There is a complaint in there, "'custom/conf/app.ini' not found".  It does exist and as far as I can tell, it's found at server start.)

Here's what shows up on the server at the same time:

```
2016/08/08 00:55:03 [T] SSH: Handshaking for 24.5.151.253:20977
2016/08/08 00:55:04 [...s/modules/ssh/ssh.go:156 func1()] [E] SearchPublicKeyByContent: public key does not exist [id: 0]
2016/08/08 00:55:04 [T] SSH: Connection from 24.5.151.253:20977 (SSH-2.0-OpenSSH_7.2)
2016/08/08 00:55:04 [T] SSH: Payload: git-receive-pack '/dave/test.git'
2016/08/08 00:55:04 [T] SSH: Arguments: [serv key-6 --config=/home/gogs/custom/conf/app.ini]
```

I uploaded my SSH public key via the web interface.  I'm able to clone using SSH, but unable to push.

I've seen other threads mention the authorized_keys file.  I'm not seeing anything added to that file.  I'm relying on gogs built-in SSH.  So I'm hoping the authorized_keys file is not needed or touched.  Is that the case?  Potentially relevant items from app.ini:

```
; Disable SSH feature when not available                                                       
DISABLE_SSH = false
; Whether use builtin SSH server or not.                                                       
START_SSH_SERVER = true
; Domain name to be exposed in clone URL                                                       
SSH_DOMAIN = %(DOMAIN)s
; Port number to be exposed in clone URL                                                       
SSH_PORT = 3022
; Port number builtin SSH server listens on                                                    
SSH_LISTEN_PORT = %(SSH_PORT)s
; Root path of SSH directory, default is '~/.ssh', but you have to use '/home/git/.ssh'.       
SSH_ROOT_PATH = /home/gogs/.ssh
; Directory to create temporary files when test publick key using ssh-keygen,                  
; default is system temporary directory.                                                       
SSH_KEY_TEST_PATH = /home/gogs/tmp
; Path to ssh-keygen, default is 'ssh-keygen' and let shell find out which one to call.        
SSH_KEYGEN_PATH = ssh-keygen
; Indicate whether to check minimum key size with corresponding type                           
MINIMUM_KEY_SIZE_CHECK = false
```

I've tried different values for SSH_ROOT_PATH.  I.e., /home/gogs/ssh (instead of .ssh).  That directory was not created.  There is a /home/gogs/data/ssh, it has only gogs' key  I don't see the uploaded key anywhere... I'm assuming it is only in the database.

Please let me know if there's any other info I can provide.  Would love to start using gogs for real.  Thanks for this amazing open source project!

**Comment by Unknwon** [comment]
Hi, I think you have never go though the install process in the page `http(s)://domain:(port)/install`.

>Please post questions or config/deploy problems on our forum: https://discuss.gogs.io, here are bugs and feature requests only.

**Comment by dncohen** [comment]
I may have been thrown off by the instructions:

```
You need to set key security -> INSTALL_LOCK to be true in file custom/conf/app.ini in order to run from source.
```

(from https://gogs.io/docs/installation/configuration_and_run)
Once I did that, there was no more /install !

I can't access my server at the moment.  I'll try it later today.  Thanks.

**Comment by Unknwon** [comment]
Anyway, the error message is very clear you did not tell which custom app.ini Gogs should use.

**Comment by dncohen** [comment]
I was running `./gogs web --config /home/gogs/custom/conf/app.ini`  And it was certainly reading that app.ini file or the SSH server would not have been running on port 3022.

Anyway, I started from scratch with a new gogs install and a new database.  I'm running with just `./gogs web`, leaving custom/conf/app.ini in its default location.  I visited /install as you suggested.  Afterward, I had to stop the server and add a few lines to /home/gogs/gogs/custom/conf/app.ini.  

```
; Whether use builtin SSH server or not.
START_SSH_SERVER = true
SSH_DOMAIN       = gogs.<my domain>
```

I don't understand exactly what /install does in addition to writing the app.ini file.  But it seems to have done the trick because I can push changes now.

I'm not a fan of keeping custom directory alongside the gogs executable.  I'm afraid I'll delete that directory when upgrading gogs.  Also, I believe for security concerns it would be best if the gogs executable can read but not write app.ini.

**Comment by Unknwon** [comment]
Does this file actually exists? `/home/gogs/custom/conf/app.ini` and `/home/gogs/gogs/custom/conf/app.ini` do not look like they have same path.

**Comment by dncohen** [comment]
I've installed gogs into /home/gogs/gogs/.  Unfortunately repetitive.  The executable is /home/gogs/gogs/gogs ;)

When I originally submitted this bug, I was using /home/gogs/custom/conf/app.ini.  My intent was to keep `custom` out of `/home/gogs/gogs` because I'll eventually replace `/home/gogs/gogs` when I upgrade.

During the course of this thread, I did a complete re-install.  When I did this, I ran /install.  It wrote `/home/gogs/gogs/custom/conf/app.ini` and I've left it there.  So, that is what I use right now.

I am starting to think, or at least hypothesize, that my gogs server setttings were correct when I submitted this issue.  But the gogs built-in ssh was misbehaving.  I am wondering when I passed in `--config ...` was that config path respected by the web server, but not being passed on to the ssh server?  When serving ssh requests does it not know about `--config` argument?
